File: javascript/smartlist/smartlist-api-mysql-project/src/modules/auth/auth.service.ts
import { Injectable } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { InjectRepository } from '@nestjs/typeorm';
import { AuthorizedToken } from 'src/entities/authorized-token.entity';
import { Candidate } from 'src/entities/candidate.entity';
import { Role } from 'src/enums/role';
import { CryptoService } from 'src/providers/crypto.service';

import { User } from '../../entities/user.entity';
import { CandidateRepository } from '../candidate/candidate.repository';
import { UserRepository } from '../user/user.repository';
import { AuthRepository } from './auth.repository';
import { AuthResponseCodes } from './auth.response.codes';
import { AuthReqDto } from './dto/request/auth.req.dto';
import { JWTPayload } from './dto/request/jwt.payload';
import { AuthRespDto } from './dto/response/auth.resp.dto';

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(AuthRepository)
    private authRepository: AuthRepository,
    @InjectRepository(UserRepository)
    private usersRepository: UserRepository,
    @InjectRepository(CandidateRepository)
    private candidateRepository: CandidateRepository,
    private cryptoService: CryptoService,
    private jwtService: JwtService,
  ) {}

  async validateUser(email: string, password: string): Promise<User> {
    const user: User = await this.usersRepository.findByEmailWithOrganization(
      email,
    );

    if (user == null) {
      throw AuthResponseCodes.USER_CREDENTIALS_INVALID;
    } else if (user.active == false) {
      throw AuthResponseCodes.USER_ACCOUNT_INACTIVE;
    }

    const isPasswordCorrect: boolean = await this.cryptoService.comparePassword(
      password,
      user.password,
    );
    if (!isPasswordCorrect) {
      throw AuthResponseCodes.USER_CREDENTIALS_INVALID;
    } else {
      return user;
    }
  }

  async validateAccessToken(payload: any) {
    const accessTokenHash = this.cryptoService.generateHash(
      JSON.stringify({
        email: payload.email,
        sub: payload.sub,
        role: payload.role,
      }),
    );

    const authorizedToken: AuthorizedToken =
      await this.authRepository.findByAccessToken(accessTokenHash);
    if (authorizedToken == null) {
      throw AuthResponseCodes.ACCESS_TOKEN_INVALID;
    }
  }

  async loginUser(authReqDto: AuthReqDto): Promise<AuthRespDto> {
    const user = await this.validateUser(authReqDto.email, authReqDto.password);

    const payload = {
      email: user.email,
      sub: user.id,
      role: user.role,
    };

    const accessToken = this.jwtService.sign(payload);
    const accessTokenHash = this.cryptoService.generateHash(
      JSON.stringify(payload),
    );

    const authorizedToken: AuthorizedToken = new AuthorizedToken();
    authorizedToken.access_token_hash = accessTokenHash;
    authorizedToken.refresh_token_hash = '';
    authorizedToken.user = user;
    await this.authRepository.deleteByUser(user);
    await this.authRepository.save(authorizedToken);

    const authRespDto: AuthRespDto = new AuthRespDto(user.id, accessToken);

    return authRespDto;
  }

  async loginCandidate(candidate: Candidate): Promise<AuthRespDto> {
    const payload = {
      email: candidate.email,
      sub: candidate.id,
      role: Role.CANDIDATE,
    };

    const accessToken = this.jwtService.sign(payload);
    const accessTokenHash = this.cryptoService.generateHash(
      JSON.stringify(payload),
    );

    const authorizedToken: AuthorizedToken = new AuthorizedToken();
    authorizedToken.access_token_hash = accessTokenHash;
    authorizedToken.refresh_token_hash = '';
    authorizedToken.candidate = candidate;
    await this.authRepository.deleteByCandidate(candidate);
    await this.authRepository.save(authorizedToken);

    const authRespDto: AuthRespDto = new AuthRespDto(candidate.id, accessToken);

    return authRespDto;
  }

  async logout(jwtPayload: JWTPayload) {
    if (jwtPayload.role == Role.CANDIDATE) {
      const candidate: Candidate = await this.candidateRepository.findById(
        jwtPayload.sub,
      );
      if (candidate != null) {
        await this.authRepository.deleteByCandidate(candidate);
      }
    } else {
      const user: User = await this.usersRepository.findById(jwtPayload.sub);
      if (user != null) {
        await this.authRepository.deleteByUser(user);
      }
    }
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/auth/AuthService.java
package com.smartlist.module.auth;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.smartlist.enums.Role;
import com.smartlist.model.AuthorizedToken;
import com.smartlist.model.Candidate;
import com.smartlist.model.User;
import com.smartlist.module.auth.dto.request.AuthReqDTO;
import com.smartlist.module.auth.dto.response.AuthRespDTO;
import com.smartlist.module.candidate.CandidateRepository;
import com.smartlist.module.user.UserRepository;
import com.smartlist.services.CryptoService;
import com.smartlist.services.JWTService;
import com.smartlist.utils.AppResponseCodes;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Optional;

@Service
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final UserRepository userRepository;
    private final AuthRepository authRepository;
    private final CandidateRepository candidateRepository;
    private final CryptoService cryptoService;
    private final JWTService jwtService;

    public AuthService(final AuthenticationManager authenticationManager, final UserRepository userRepository, final AuthRepository authRepository,
                       final CandidateRepository candidateRepository, final CryptoService cryptoService,
                       final JWTService jwtService){
        this.authenticationManager = authenticationManager;
        this.userRepository = userRepository;
        this.authRepository = authRepository;
        this.candidateRepository = candidateRepository;
        this.cryptoService = cryptoService;
        this.jwtService = jwtService;
    }


    @Transactional
    public AuthRespDTO loginUser(final AuthReqDTO authReqDTO) {
        try {
            Authentication authenticate = this.authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(
                    authReqDTO.getEmail(), authReqDTO.getPassword()
            ));

            User user = (User) authenticate.getPrincipal();

            String accessToken = this.jwtService.buildJWTToken(user);
            String accessTokenHash = "";
            try {
                accessTokenHash = this.cryptoService.generateHash(accessToken);
            }catch (NoSuchAlgorithmException ex){
                throw AppResponseCodes.BAD_REQUEST;
            }

            AuthorizedToken authorizedToken = new AuthorizedToken();
            authorizedToken.setAccessTokenHash(accessTokenHash);
            authorizedToken.setRefreshTokenHash("");
            authorizedToken.setUser(user);
            this.authRepository.deleteByUser(user);
            this.authRepository.save(authorizedToken);

            return new AuthRespDTO(user.getId(), accessToken);

        }catch (BadCredentialsException ex) {
            throw AppResponseCodes.UNAUTHORISED;
        }

    }

    @Transactional
    public AuthRespDTO loginCandidate(final Candidate candidate) {
        HashMap<String, String> claims = new HashMap<>();
        claims.put("id", String.valueOf(candidate.getId()));
        claims.put("role", Role.CANDIDATE.name());

        ObjectMapper objectMapper = new ObjectMapper();

        String accessToken = this.jwtService.generateToken(candidate.getEmail(), claims);
        String accessTokenHash ="";
        try {
            accessTokenHash = this.cryptoService.generateHash(objectMapper.writeValueAsString(claims));
        }catch (JsonProcessingException | NoSuchAlgorithmException ex){
            throw AppResponseCodes.BAD_REQUEST;
        }


        AuthorizedToken authorizedToken  = new AuthorizedToken();
        authorizedToken.setAccessTokenHash(accessTokenHash);
        authorizedToken.setRefreshTokenHash("");
        authorizedToken.setCandidate(candidate);

        this.authRepository.deleteByCandidate(candidate);
        this.authRepository.save(authorizedToken);

        return new AuthRespDTO(candidate.getId(), accessToken);

    }

    @Transactional
    public void logout(final HttpServletRequest request) {
        String jwt = this.jwtService.getBearerTokenFromHeader(request.getHeader(HttpHeaders.AUTHORIZATION));
        if (this.jwtService.getRole(jwt).equals(Role.CANDIDATE)) {
            Optional<Candidate> optionalCandidate = this.candidateRepository.findByEmail(this.jwtService.extractUsername(jwt));
            optionalCandidate.ifPresent(this.authRepository::deleteByCandidate);
        } else {
            Optional<User> optionalUser = this.userRepository.findByEmail(this.jwtService.extractUsername(jwt));
            optionalUser.ifPresent(this.authRepository::deleteByUser);
        }
    }

    private ResponseStatusException throwIfUserNotDefined()  {
        return AuthResponseCodes.USER_CREDENTIALS_INVALID;
    }


}
